components/plan-build/bin/hab-plan-build.sh

```diff
@@ -870,6 +870,19 @@ _return_or_append_to_set() {
   return 0
 }
 
+# **Internal** Appends an entry to the given array and removes any entries
+# already in the array that match the passed entry.
```

Comment:
This function is fine based on the current call sites, but I think we might want to add more detail to the comment or change the same to indicate how to use it safely since it has some confusing behavior when used on more arbitrary input.

The opportunity for misuse is produced by

```
   dedupe_set=("${deps[@]/$1}")
```

This is doing string replacement on each member of the array, replacing any matched substring with the empty string.  This produces two problems:

1) For exact matches, it produces an empty element in the array, thus this function ends up extending the length of the array by 1 empyt element.

```
> _dedupe_and_append_to_set_debug bar bar bat baz
            deps: bar bat baz
        deps len: 3
      dedupe_set:  bat baz
  dedupe_set len: 3
    appended_set:  bat baz bar
appended_set len: 4
```

at the callsites this is OK because of our use of `read` to rebuild the array.

2) If someone where to use this on a set with the possibility of non-exact matches, you'd get very surprising results:

```
> _dedupe_and_append_to_set_debug b bar bat baz
            deps: bar bat baz
        deps len: 3
      dedupe_set: ar at az
  dedupe_set len: 3
    appended_set: ar at az b
appended_set len: 4
```

At our callsites this is currently OK because we are using this on idents that should never have substring matches.

Reply:
thanks for this clarification @stevendanna ! I was a bit nervous about this function since bash is a second shell language for me. I renamed the function to be specific to our use case and not a generic array utility.